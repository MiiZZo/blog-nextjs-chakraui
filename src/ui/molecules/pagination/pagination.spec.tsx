import React from "react";
import { cleanup, fireEvent } from "@testing-library/react";
import { myRender } from "@lib/testing/render";
import Pagination from "./pagination";

afterEach(cleanup);

describe("pagination component", () => {
  const defaultCurrent = 1;
  const total = 10;
  const onChange = jest.fn();

  it("numbered buttons should not be more than total", async () => {
    const current = 18;
    const total = 20;
    const { getAllByTestId } = myRender(
      <Pagination total={total} current={current} onChange={onChange} />
    );
    const paginateButtons = await getAllByTestId("paginate-button");
    const paginateButtonsCount = paginateButtons.length;

    expect(paginateButtons[paginateButtonsCount - 1].textContent).toBe(
      total.toString()
    );
  });

  it("onChange function should be called 3 times", async () => {
    const onChange = jest.fn();
    const current = 10;
    const total = 20;
    const { getByTestId, getAllByTestId } = myRender(
      <Pagination current={current} total={total} onChange={onChange} />
    );

    const firstPaginateButton = getByTestId("first-paginate-button");
    const lastPaginateButton = getByTestId("last-paginate-button");
    const paginateButton = await getAllByTestId("paginate-button")[0];

    fireEvent.click(firstPaginateButton);
    fireEvent.click(lastPaginateButton);
    fireEvent.click(paginateButton);

    expect(onChange).toBeCalledTimes(3);
  });

  it("last paginate button should not be disabled", () => {
    const { getByTestId } = myRender(
      <Pagination current={defaultCurrent} total={total} onChange={onChange} />
    );
    const lastPaginateButton = getByTestId("last-paginate-button");

    expect(lastPaginateButton).not.toBeDisabled();
  });

  it("first paginate button should be disabled", async () => {
    const { getByTestId } = myRender(
      <Pagination current={defaultCurrent} total={total} onChange={onChange} />
    );
    const firstPaginateButton = getByTestId("first-paginate-button");

    expect(firstPaginateButton).toBeDisabled();
  });

  it("should contains five paginate button", async () => {
    const { findAllByTestId } = myRender(
      <Pagination current={defaultCurrent} total={total} onChange={onChange} />
    );

    const elementsContainer = await findAllByTestId("paginate-button");

    expect(elementsContainer.length).toBe(5);
  });

  it("should match a shapshot", () => {
    const { asFragment } = myRender(
      <Pagination current={defaultCurrent} total={total} onChange={onChange} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
