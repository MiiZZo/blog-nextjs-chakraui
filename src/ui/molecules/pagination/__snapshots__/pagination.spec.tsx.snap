// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pagination component should match a shapshot 1`] = `
<DocumentFragment>
  <div
    class="css-8xl60i"
  >
    <div
      class="css-u4p24i"
    >
      <button
        aria-disabled="true"
        aria-label="move to first page"
        class="css-1mrta8r"
        data-testid="first-paginate-button"
        disabled=""
        type="button"
      >
        <svg
          aria-hidden="true"
          class="css-1im46kq"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <g
            fill="currentColor"
          >
            <path
              d="M10.416,12a2.643,2.643,0,0,1,.775-1.875L20.732.584a1.768,1.768,0,0,1,2.5,2.5l-8.739,8.739a.25.25,0,0,0,0,.354l8.739,8.739a1.768,1.768,0,0,1-2.5,2.5l-9.541-9.541A2.643,2.643,0,0,1,10.416,12Z"
            />
            <path
              d="M.25,12a2.643,2.643,0,0,1,.775-1.875L10.566.584a1.768,1.768,0,0,1,2.5,2.5L4.327,11.823a.25.25,0,0,0,0,.354l8.739,8.739a1.768,1.768,0,0,1-2.5,2.5L1.025,13.875A2.643,2.643,0,0,1,.25,12Z"
            />
          </g>
        </svg>
      </button>
      <div
        class="css-sarrtm"
      >
        <button
          class="css-wcven5"
          data-testid="paginate-button"
          type="button"
        >
          1
        </button>
        <button
          class="css-mtrgpb"
          data-testid="paginate-button"
          type="button"
        >
          2
        </button>
        <button
          class="css-mtrgpb"
          data-testid="paginate-button"
          type="button"
        >
          3
        </button>
        <button
          class="css-mtrgpb"
          data-testid="paginate-button"
          type="button"
        >
          4
        </button>
        <button
          class="css-ej70pr"
          data-testid="paginate-button"
          type="button"
        >
          5
        </button>
      </div>
      <button
        aria-label="move to last page"
        class="css-1wf4v31"
        data-testid="last-paginate-button"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="css-1im46kq"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <g
            fill="currentColor"
          >
            <path
              d="M13.584,12a2.643,2.643,0,0,1-.775,1.875L3.268,23.416a1.768,1.768,0,0,1-2.5-2.5l8.739-8.739a.25.25,0,0,0,0-.354L.768,3.084a1.768,1.768,0,0,1,2.5-2.5l9.541,9.541A2.643,2.643,0,0,1,13.584,12Z"
            />
            <path
              d="M23.75,12a2.643,2.643,0,0,1-.775,1.875l-9.541,9.541a1.768,1.768,0,0,1-2.5-2.5l8.739-8.739a.25.25,0,0,0,0-.354L10.934,3.084a1.768,1.768,0,0,1,2.5-2.5l9.541,9.541A2.643,2.643,0,0,1,23.75,12Z"
            />
          </g>
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
